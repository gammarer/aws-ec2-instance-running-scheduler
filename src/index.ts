import * as crypto from 'crypto';
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as scheduler from 'aws-cdk-lib/aws-scheduler';
import { Construct } from 'constructs';

export interface ScheduleProperty {
  readonly timezone: string;
  readonly minute?: string;
  readonly hour?: string;
  readonly week?: string;
}

export interface TargetsProperty {
  readonly instances: string[];
  readonly stopSchedule: ScheduleProperty;
  readonly startSchedule: ScheduleProperty;
}

export interface Ec2InstanceRunningSchedulerProps {
  readonly targets: TargetsProperty[];
}

export class Ec2InstanceRunningScheduler extends Construct {
  constructor(scope: Construct, id: string, props: Ec2InstanceRunningSchedulerProps) {
    super(scope, id);

    const account = cdk.Stack.of(this).account;
    //const stackName: string = cdk.Stack.of(this).stackName;
    //const region = cdk.Stack.of(this).region;

    const randomNameKey = crypto.createHash('shake256', { outputLength: 4 })
      .update(`${cdk.Names.uniqueId(scope)}-${cdk.Names.uniqueId(this)}`)
      .digest('hex');

    // ðŸ‘‡EventBridge Scheduler IAM Role
    const schedulerExecutionRole = new iam.Role(this, 'SchedulerExecutionRole', {
      roleName: `stop-ec2-instance-schedule-${randomNameKey}-exec-role`,
      assumedBy: new iam.ServicePrincipal('scheduler.amazonaws.com'),
      inlinePolicies: {
        ['ec2-instance-stop-policy']: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                'ec2:StartInstances',
                'ec2:StopInstances',
              ],
              resources: [
                `arn:aws:ec2:*:${account}:instance/*`,
              ],
            }),
          ],
        }),
      },
    });

    // Object.entries(props.targets)

    for (const target of props.targets) {
      // ðŸ‘‡Create random string
      const scheduleNameKey = crypto.createHash('shake256', { outputLength: 4 })
        .update(target.instances.toString())
        .digest('hex');

      // ðŸ‘‡EventBridge Scheduler for DB Instance Stop
      new scheduler.CfnSchedule(this, `StopSchedule${scheduleNameKey.toUpperCase()}`, {
        name: `auto-stop-ec2-instance-${scheduleNameKey}-schedule`,
        description: `auto stop ec2 instance (${target.instances.join(',')}) schedule.`,
        state: 'ENABLED',
        //groupName: scheduleGroup.name, // default
        flexibleTimeWindow: {
          mode: 'OFF',
        },
        scheduleExpressionTimezone: target.stopSchedule.timezone,
        scheduleExpression: (() => {
          // default: weekday 21:10
          const minute: string = target.stopSchedule.minute ?? '10';
          const hour: string = target.stopSchedule.hour ?? '21';
          const week: string = target.stopSchedule.week ?? 'MON-FRI';
          return `cron(${minute} ${hour} ? * ${week} *)`;
        })(),
        target: {
          arn: 'arn:aws:scheduler:::aws-sdk:ec2:stopInstances',
          roleArn: schedulerExecutionRole.roleArn,
          input: JSON.stringify({ InstanceIds: target.instances }),
          retryPolicy: {
            maximumEventAgeInSeconds: 60,
            maximumRetryAttempts: 0,
          },
        },
      });

      // ðŸ‘‡EventBridge Scheduler for DB Instance Start
      new scheduler.CfnSchedule(this, `StartSchedule${scheduleNameKey.toUpperCase()}`, {
        name: `auto-start-ec2-instance-${scheduleNameKey}-schedule`,
        description: `auto start ec2 instance(${target.instances.join(',')}) schedule.`,
        state: 'ENABLED',
        //groupName: scheduleGroup.name, // default
        flexibleTimeWindow: {
          mode: 'OFF',
        },
        scheduleExpressionTimezone: target.startSchedule.timezone,
        scheduleExpression: (() => {
          // default: weekday 07:50
          const minute: string = target.startSchedule.minute ?? '50';
          const hour: string = target.startSchedule.hour ?? '7';
          const week: string = target.startSchedule.week ?? 'MON-FRI';
          return `cron(${minute} ${hour} ? * ${week} *)`;
        })(),
        target: {
          arn: 'arn:aws:scheduler:::aws-sdk:ec2:startInstances',
          roleArn: schedulerExecutionRole.roleArn,
          input: JSON.stringify({ InstanceIds: target.instances }),
          retryPolicy: {
            maximumEventAgeInSeconds: 60,
            maximumRetryAttempts: 0,
          },
        },
      });
    }
  }
}
