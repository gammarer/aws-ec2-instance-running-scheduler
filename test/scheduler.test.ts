import { App, Stack } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { Ec2InstanceRunningScheduler } from '../src';

describe('Ec2InstanceRunningScheduler Type=Cluster Testing', () => {

  describe('default schedule', () => {
    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    new Ec2InstanceRunningScheduler(stack, 'Ec2InstanceRunningScheduler', {
      targets: [
        {
          instances: ['i-0af01c0123456789a', 'i-0af01c0123456789b'],
          startSchedule: {
            timezone: 'UTC',
          },
          stopSchedule: {
            timezone: 'UTC',
          },
        },
      ],
    });

    const template = Template.fromStack(stack);

    // Schedule execution role
    it('Should have schedule execution role', async () => {
      template.hasResourceProperties('AWS::IAM::Role', Match.objectEquals({
        RoleName: Match.stringLikeRegexp('stop-ec2-instance-schedule-.*-exec-role'),
        AssumeRolePolicyDocument: Match.objectEquals({
          Version: '2012-10-17',
          Statement: Match.arrayWith([
            Match.objectEquals({
              Effect: 'Allow',
              Principal: {
                Service: 'scheduler.amazonaws.com',
              },
              Action: 'sts:AssumeRole',
            }),
          ]),
        }),
        Policies: Match.arrayEquals([
          {
            PolicyName: 'ec2-instance-stop-policy',
            PolicyDocument: Match.objectEquals({
              Version: '2012-10-17',
              Statement: [
                Match.objectEquals({
                  Effect: 'Allow',
                  Action: Match.arrayWith([
                    'ec2:StartInstances',
                    'ec2:StopInstances',
                  ]),
                  Resource: 'arn:aws:ec2:*:123456789012:instance/*',
                }),
              ],
            }),
          },
        ]),
      }));
    });

    // Start Schedule testing
    it('Should have Start Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-start-ec2-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(50 7 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:ec2:startInstances',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2InstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({ InstanceIds: ['i-0af01c0123456789a', 'i-0af01c0123456789b'] }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    // Stop Schedule testing
    it('Should have Stop Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-stop-ec2-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'UTC',
        ScheduleExpression: 'cron(10 21 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:ec2:stopInstances',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2InstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({ InstanceIds: ['i-0af01c0123456789a', 'i-0af01c0123456789b'] }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot('default');
    });

  });

  describe('specify schedule', () => {
    const app = new App();
    const stack = new Stack(app, 'TestingStack', {
      env: {
        account: '123456789012',
        region: 'us-east-1',
      },
    });

    new Ec2InstanceRunningScheduler(stack, 'Ec2InstanceRunningScheduler', {
      targets: [
        {
          instances: ['i-0af01c0123456789a', 'i-0af01c0123456789b'],
          startSchedule: {
            timezone: 'Asia/Tokyo',
            minute: '55',
            hour: '8',
            week: 'MON-FRI',
          },
          stopSchedule: {
            timezone: 'Asia/Tokyo',
            minute: '5',
            hour: '19',
            week: 'MON-FRI',
          },
        },
      ],
    });

    const template = Template.fromStack(stack);

    // Start Schedule testing
    it('Should have Start Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-start-ec2-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'Asia/Tokyo',
        ScheduleExpression: 'cron(55 8 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:ec2:startInstances',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2InstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({ InstanceIds: ['i-0af01c0123456789a', 'i-0af01c0123456789b'] }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    // Stop Schedule testing
    it('Should have Stop Schedule', async () => {
      template.hasResourceProperties('AWS::Scheduler::Schedule', Match.objectEquals({
        Name: Match.stringLikeRegexp('auto-stop-ec2-instance-.*-schedule'),
        Description: Match.anyValue(),
        State: 'ENABLED',
        FlexibleTimeWindow: {
          Mode: 'OFF',
        },
        ScheduleExpressionTimezone: 'Asia/Tokyo',
        ScheduleExpression: 'cron(5 19 ? * MON-FRI *)',
        Target: Match.objectEquals({
          Arn: 'arn:aws:scheduler:::aws-sdk:ec2:stopInstances',
          RoleArn: {
            'Fn::GetAtt': [
              Match.stringLikeRegexp('Ec2InstanceRunningSchedulerSchedulerExecutionRole.*'),
              'Arn',
            ],
          },
          Input: JSON.stringify({ InstanceIds: ['i-0af01c0123456789a', 'i-0af01c0123456789b'] }),
          RetryPolicy: {
            MaximumEventAgeInSeconds: 60,
            MaximumRetryAttempts: 0,
          },
        }),
      }));
    });

    it('Should match snapshot', async () => {
      expect(template.toJSON()).toMatchSnapshot('specify');
    });

  });
});