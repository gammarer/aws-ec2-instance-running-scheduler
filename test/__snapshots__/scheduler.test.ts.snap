// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ec2InstanceRunningScheduler Type=Cluster Testing default schedule Should match snapshot: default 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Ec2InstanceRunningSchedulerSchedulerExecutionRole16E4E2DD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "ec2:StartInstances",
                    "ec2:StopInstances",
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:ec2:*:123456789012:instance:*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ec2-instance-stop-policy",
          },
        ],
        "RoleName": "stop-ec2-instance-schedule-f180f971-exec-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "Ec2InstanceRunningSchedulerStartSchedule1FDB0719F41BCB51": Object {
      "Properties": Object {
        "Description": "auto start ec2 instance(i-0af01c0123456789a,i-0af01c0123456789b) schedule.",
        "FlexibleTimeWindow": Object {
          "Mode": "OFF",
        },
        "Name": "auto-start-ec2-instance-1fdb0719-schedule",
        "ScheduleExpression": "cron(50 7 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": Object {
          "Arn": "arn:aws:scheduler:::aws-sdk:ec2:startInstances",
          "Input": "{\\"InstanceIds\\":[\\"i-0af01c0123456789a\\",\\"i-0af01c0123456789b\\"]}",
          "RetryPolicy": Object {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": Object {
            "Fn::GetAtt": Array [
              "Ec2InstanceRunningSchedulerSchedulerExecutionRole16E4E2DD",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "Ec2InstanceRunningSchedulerStopSchedule1FDB07199E70AFE7": Object {
      "Properties": Object {
        "Description": "auto stop ec2 instance (i-0af01c0123456789a,i-0af01c0123456789b) schedule.",
        "FlexibleTimeWindow": Object {
          "Mode": "OFF",
        },
        "Name": "auto-stop-ec2-instance-1fdb0719-schedule",
        "ScheduleExpression": "cron(10 21 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "UTC",
        "State": "ENABLED",
        "Target": Object {
          "Arn": "arn:aws:scheduler:::aws-sdk:ec2:stopInstances",
          "Input": "{\\"InstanceIds\\":[\\"i-0af01c0123456789a\\",\\"i-0af01c0123456789b\\"]}",
          "RetryPolicy": Object {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": Object {
            "Fn::GetAtt": Array [
              "Ec2InstanceRunningSchedulerSchedulerExecutionRole16E4E2DD",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Ec2InstanceRunningScheduler Type=Cluster Testing specify schedule Should match snapshot: specify 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "Ec2InstanceRunningSchedulerSchedulerExecutionRole16E4E2DD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "ec2:StartInstances",
                    "ec2:StopInstances",
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:ec2:*:123456789012:instance:*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ec2-instance-stop-policy",
          },
        ],
        "RoleName": "stop-ec2-instance-schedule-f180f971-exec-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "Ec2InstanceRunningSchedulerStartSchedule1FDB0719F41BCB51": Object {
      "Properties": Object {
        "Description": "auto start ec2 instance(i-0af01c0123456789a,i-0af01c0123456789b) schedule.",
        "FlexibleTimeWindow": Object {
          "Mode": "OFF",
        },
        "Name": "auto-start-ec2-instance-1fdb0719-schedule",
        "ScheduleExpression": "cron(55 8 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": Object {
          "Arn": "arn:aws:scheduler:::aws-sdk:ec2:startInstances",
          "Input": "{\\"InstanceIds\\":[\\"i-0af01c0123456789a\\",\\"i-0af01c0123456789b\\"]}",
          "RetryPolicy": Object {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": Object {
            "Fn::GetAtt": Array [
              "Ec2InstanceRunningSchedulerSchedulerExecutionRole16E4E2DD",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
    "Ec2InstanceRunningSchedulerStopSchedule1FDB07199E70AFE7": Object {
      "Properties": Object {
        "Description": "auto stop ec2 instance (i-0af01c0123456789a,i-0af01c0123456789b) schedule.",
        "FlexibleTimeWindow": Object {
          "Mode": "OFF",
        },
        "Name": "auto-stop-ec2-instance-1fdb0719-schedule",
        "ScheduleExpression": "cron(5 19 ? * MON-FRI *)",
        "ScheduleExpressionTimezone": "Asia/Tokyo",
        "State": "ENABLED",
        "Target": Object {
          "Arn": "arn:aws:scheduler:::aws-sdk:ec2:stopInstances",
          "Input": "{\\"InstanceIds\\":[\\"i-0af01c0123456789a\\",\\"i-0af01c0123456789b\\"]}",
          "RetryPolicy": Object {
            "MaximumEventAgeInSeconds": 60,
            "MaximumRetryAttempts": 0,
          },
          "RoleArn": Object {
            "Fn::GetAtt": Array [
              "Ec2InstanceRunningSchedulerSchedulerExecutionRole16E4E2DD",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Scheduler::Schedule",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
